plugins/module_utils/foreman_helper.py

```diff
@@ -471,6 +465,77 @@ def find_puppetclass(self, name, environment=None, params=None, failsafe=False,
+
+    def lookup_entity(self, key):
+        if key not in self.foreman_params:
+            return None
+
+        entity_spec = self.foreman_spec[key]
+        if entity_spec.get('resolved') or entity_spec.get('type') not in ('entity', 'entity_list'):
+            # Already looked up or not an entity(_list) so nothing to do
+            return self.foreman_params[key]
+
+        resource_type = entity_spec['resource_type']
+        failsafe = entity_spec.get('failsafe', False)
+        thin = entity_spec.get('thin', True)
+        params = {}
+        try:
+            if 'scope' in entity_spec:
+                for scope in entity_spec['scope']:
+                    params.update(self.scope_for(scope))
+        except TypeError:
+            if failsafe:
+                if entity_spec.get('type') == 'entity':
+                    result = None
+                else:
+                    result = [None for value in self.foreman_params[key]]
+            else:
+                self.fail_json(msg="Failed to lookup scope {0} while searching for {1}.".format(entity_spec['scope'], resource_type))
+        else:
+            # No exception happend => scope is in place
+            if resource_type == 'operatingsystems':
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_operatingsystem(self.foreman_params[key], params=params, failsafe=failsafe, thin=thin)
+                else:
+                    result = [self.find_operatingsystem(value, params=params, failsafe=failsafe, thin=thin) for value in self.foreman_params[key]]
+            elif resource_type == 'puppetclasses':
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_puppetclass(self.foreman_params[key], params=params, failsafe=failsafe, thin=thin)
+                else:
+                    result = [self.find_puppetclass(value, params=params, failsafe=failsafe, thin=thin) for value in self.foreman_params[key]]
+            else:
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_resource_by(
+                        resource=resource_type,
+                        value=self.foreman_params[key],
+                        search_field=entity_spec.get('search_by', ENTITY_KEYS.get(resource_type, 'name')),
+                        search_operator=entity_spec.get('search_operator', '='),
+                        failsafe=failsafe, thin=thin, params=params,
+                    )
+                else:
+                    result = [self.find_resource_by(
+                        resource=resource_type,
+                        value=value,
+                        search_field=entity_spec.get('search_by', ENTITY_KEYS.get(resource_type, 'name')),
+                        search_operator=entity_spec.get('search_operator', '='),
+                        failsafe=failsafe, thin=thin, params=params,
+                    ) for value in self.foreman_params[key]]
```

Comment:
I find the "handle entity vs entity_list" in this block very repetitive, but don't have a nice solution for it just now. Maybe something we can improve in the future, tho.

Reply:
Maybe by embedding the `if resource_type == 'operatingsystems':` into `find_resource_by`.
But i think the operatingsystem module needs an unenhanced version of the latter.